Quarkus ArC integration refactoring (#13749)
quarkusio/quarkus: docs/src/main/asciidoc/cdi-integration.adoc

////
This guide is maintained in the main Quarkus repository
and pull requests should be submitted there:
https://github.com/quarkusio/quarkus/tree/master/docs/src/main/asciidoc
////
= Quarkus - CDI Integration Guide

include::./attributes.adoc[]
:numbered:
:sectnums:
:sectnumlevels: 4
:toc:
:toclevels: 2

The ArC container is bootstrapped at build time.
The downside of this approach is that CDI Portable Extensions cannot be supported.
Nevertheless, the functionality can be achieved using the Quarkus-specific extensions API.

The container is bootstrapped in multiple phases.
From a high level perspective these phases go as follows:

1. Initialization
2. Bean discovery 
3. Registration of synthetic components
4. Validation

In the _initialization_ phase the preparatory work is being carried out and custom contexts are registered.
_Bean discovery_ is then the process where the container analyzes all application classes, identifies beans and wires them all together based on the provided metadata.
Subsequently, the extensions can register _synthetic components_.

> Attributes of these components are fully contolled by the extensions, i.e. are not derived from an existing class.

Reviewer: ```suggestion
Attributes of these components are fully controlled by the extensions, i.e. are not derived from an existing class.
```

It would also be helpful to add an example of what is meant by "not derived from an existing class" to make synthetic beans more palatable to users. In particular, when they make sense as opposed to "regular" beans…

Finally, the _deployment is validated_.
For example, the container validates every injection point in the application and fails the build if there is no bean that satisfies the given required type and qualifiers.

TIP: You can see more information about the bootstrap by enabling additional logging. Simply run the Maven build with `-X,--debug` and grep lines that contain `io.quarkus.arc`. In the <<cdi-reference.adoc#dev-mode,development mode>>, you can use `quarkus.log.category."io.quarkus.arc.processor".level=DEBUG` and two special endpoints are also registered automatically to provide some basic debug info in the JSON format.

Quarkus build steps can produce and consume various build items and hook into each phase.
In the following sections we will describe all the relevant build items and common scenarios.

[IMPORTANT]
.Metadata Sources
====
Classes and annotations are the primary source of bean-level metadata.
However, extensions can add, remove or transform the metadata at certain stages of the bootstrap.
Moreover, extensions can also register <<synthetic_beans,synthetic components>>. 

> This is an important aspect to realize when integrating CDI components in Quarkus.

Reviewer: It would be helpful to explain in greater details why it's important (because some beans that you might expect might not be there anymore at the end of the process, more are added, etc.) and maybe provide some examples to show concretely what might happen.

Author: That's a good idea but I'd like to keep this part of the docs concise. What about this: _"Quarkus extensions can turn classes, that would be otherwise ignored, into beans and vice versa. For example, a class that declares a `@Scheduled` method is always registered as a bean even if it is not annotated with a bean defining annotation and would be normally ignored."_

In fact, it's probably a good topic for a blogpost or something like that. The list of examples of "what might happen" could be very long...

Reviewer: I agree that it would be a good material for a blog post (and I intend to document some of my extension writing experience in such blogs). 
However, when looking at reference doc, I'd also expect it to be as complete as possible so that I don't have to look at lots of different places to find the information I need. The example you give would probably be good enough if we don't want to get into too many details here.

Author: > I'd also expect it to be as complete as possible so that I don't have to look at lots of different places to find the information I need.

It's always a question of balance. My concern is that "too exhaustive" docs may scare people ;-).

We can always add more examples.

====

== Use Case - My Class Is Not Recognized as a Bean

An `UnsatisfiedResolutionException` indicates a problem during <<cdi.adoc#typesafe_resolution,typesafe resolution>>.

> Sometimes an injection point cannot be satisifed even if there is a class on the class path that appears to be eligible for injection.

Reviewer: ```suggestion
Sometimes an injection point cannot be satisfied even if there is a class on the class path that appears to be eligible for injection.
```

////
This guide is maintained in the main Quarkus repository
and pull requests should be submitted there:
https://github.com/quarkusio/quarkus/tree/master/docs/src/main/asciidoc
////
= Quarkus - CDI Integration Guide

include::./attributes.adoc[]
:numbered:
:sectnums:
:sectnumlevels: 4
:toc:
:toclevels: 2

> The ArC container is bootstrapped at build time.

Author: I don't think we have a "strong" policy, i.e. we already mention ArC in the [ref guide](https://quarkus.io/guides/cdi-reference#arc-configuration-reference). But I like the wording "ArC, the CDI container,..."

      }

      public void transform(TransformationContext context) {
         if (context.getTarget().asClass().name().toString().equals("org.acme.Bar")) {
            context.transform().add(MyInterceptorBinding.class).done(); <2>
         }
      }
    });
}
----
<1> The transformer is only applied to classes.
<2> If the class name equals to `org.acme.Bar` then add `@MyInterceptorBinding`. Don't forget to invoke `Transformation#done()`.

NOTE: Keep in mind that annotation transformers must be produced _before_ the bean discovery starts.

Build steps can query the transformed annotations for a given annotation target via the `TransformedAnnotationsBuildItem`.

.`TransformedAnnotationsBuildItem` Example
[source,java]
----
@BuildStep
void queryAnnotations(TransformedAnnotationsBuildItem transformedAnnotations, BuildProducer<MyBuildItem> myBuildItem) {
   ClassInfo myClazz = ...;
   if (transformedAnnotations.getAnnotations(myClazz).isEmpty()) { <1>
     myBuildItem.produce(new MyBuildItem(););
   }
}
----
<1> `TransformedAnnotationsBuildItem.getAnnotations()` will return a possibly transformed set of annotations.

> NOTE: There are other build items specialized on transformation: <<additional_interceptor_bindings>> and <<injection_point_transformation>>.

Reviewer 2: ```suggestion
NOTE: There are other build items specialized in transformation: <<additional_interceptor_bindings>> and <<injection_point_transformation>>.
```

The output of a bean configurator is recorded as bytecode. 
Therefore, there are some limitations in how a synthetic bean instance is created at runtime. 
You can:

1. Generate the bytecode directly via `ExtendedBeanConfigurator.creator(Consumer<MethodCreator>)`.
2. Pass a `io.quarkus.arc.BeanCreator` implementation via `ExtendedBeanConfigurator#creator(Class<? extends BeanCreator<U>>)`, and possibly specify some parameters via `ExtendedBeanConfigurator#param()`.
3. Produce the runtime instance through a proxy returned from a <<writing-extensions.adoc#bytecode-recording,`@Recorder`>> method and set it via `ExtendedBeanConfigurator#runtimeValue(RuntimeValue<?>)` or `ExtendedBeanConfigurator#supplier(Supplier<?>)`

.`SyntheticBeanBuildItem` Example 2
[source,java]
----
@BuildStep
@Record(STATIC_INIT) <1>
SyntheticBeanBuildItem syntheticBean(TestRecorder recorder) {
   return SyntheticBeanBuildItem.configure(Foo.class).scope(Singleton.class)
                .runtimeValue(recorder.createFoo()) <2>
                .done();
}
----
<1> By default, a synthetic bean is initialized during `STATIC_INIT`.
<2> The bean instance is supplied by a value returned from a recorder method.

> It is possible to mark a synthetic bean to be initialized during `RUNTIME_INIT`:

Reviewer 2: Maybe add a link to other Quarkus docs explaining difference between the two?

include::./attributes.adoc[]
:numbered:
:sectnums:
:sectnumlevels: 4
:toc:
:toclevels: 2

ArC, the CDI container, is bootstrapped at build time.
The downside of this approach is that CDI Portable Extensions cannot be supported.
Nevertheless, the functionality can be achieved using the Quarkus-specific extensions API.

The container is bootstrapped in multiple phases.
From a high level perspective these phases go as follows:

1. Initialization
2. Bean discovery 
3. Registration of synthetic components
4. Validation

In the _initialization_ phase the preparatory work is being carried out and custom contexts are registered.
_Bean discovery_ is then the process where the container analyzes all application classes, identifies beans and wires them all together based on the provided metadata.
Subsequently, the extensions can register _synthetic components_.
Attributes of these components are fully controlled by the extensions, i.e. are not derived from an existing class.
Finally, the _deployment is validated_.
For example, the container validates every injection point in the application and fails the build if there is no bean that satisfies the given required type and qualifiers.

> TIP: You can see more information about the bootstrap by enabling additional logging. Simply run the Maven build with `-X,--debug` and grep lines that contain `io.quarkus.arc`. In the <<cdi-reference.adoc#dev-mode,development mode>>, you can use `quarkus.log.category."io.quarkus.arc.processor".level=DEBUG` and two special endpoints are also registered automatically to provide some basic debug info in the JSON format.

Reviewer 3: I'm surprised by the comma in `-X,--debug` ?

Reviewer 2: Those should be two different options to achieve the same logging. You can use `-X` or `-debug` or `--debug` to achieve the same thing.

Author: It should read `-X` or `--debug`.

Quarkus build steps can produce and consume various build items and hook into each phase.
In the following sections we will describe all the relevant build items and common scenarios.

[IMPORTANT]
.Metadata Sources
====
Classes and annotations are the primary source of bean-level metadata.
However, extensions can add, remove or transform the metadata at certain stages of the bootstrap.
Moreover, extensions can also register <<synthetic_beans,synthetic components>>. 
This is an important aspect to realize when integrating CDI components in Quarkus.

This way, extensions can turn classes, that would be otherwise ignored, into beans and vice versa.
For example, a class that declares a `@Scheduled` method is always registered as a bean even if it is not annotated with a bean defining annotation and would be normally ignored.
====

== Use Case - My Class Is Not Recognized as a Bean

An `UnsatisfiedResolutionException` indicates a problem during <<cdi.adoc#typesafe_resolution,typesafe resolution>>.

> Sometimes an injection point cannot be satisfied even if there is a class on the class path that appears to be eligible for injection.

Reviewer 3: ```suggestion
Sometimes an injection point cannot be satisfied even if there is a class on the classpath that appears to be eligible for injection.
```

     return new AdditionalBeanBuildItem(SmallRyeHealthReporter.class, HealthServlet.class)); <1>
}
----
<1> `AdditionalBeanBuildItem.Builder` can be used for more complex use cases.

Bean classes added via `AdditionalBeanBuildItem` are _removable_ by default.
If the container considers them <<cdi-reference.adoc#remove_unused_beans,unused>>, they are just ignored.
However, you can use `AdditionalBeanBuildItem.Builder.setUnremovable()` method to instruct the container to never remove bean classes registered via this build item.
See also <<cdi-reference.adoc#remove_unused_beans,Removing Unused Beans>> and <<unremovable_builditem>> for more details.

It is aso possible to set the default scope via `AdditionalBeanBuildItem.Builder#setDefaultScope()`.
The default scope is only used if there is no scope declared on the bean class. 

NOTE: If no default scope is specified the `@Dependent` pseudo-scope is used.

=== _Reason 2_: Class Is Discovered but Has No Bean Defining Annotation

In Quarkus, the application is represented by a single bean archive with the https://docs.jboss.org/cdi/spec/2.0/cdi-spec.html#default_bean_discovery[bean discovery mode `annotated`, window="_blank"].

> Therefore, bean classes that don't have a http://docs.jboss.org/cdi/spec/2.0/cdi-spec.html#bean_defining_annotations[bean defining annotation, window="_blank"] are ignored.

Reviewer 3: ```suggestion
Therefore, bean classes that don't have a https://docs.jboss.org/cdi/spec/2.0/cdi-spec.html#bean_defining_annotations[bean defining annotation, window="_blank"] are ignored.
```

Bean defining annotations are declared on the class-level and incluce scopes, stereotypes and `@Interceptor`.

_Solution 1_: Use the `AutoAddScopeBuildItem`. This build item can be used to add a scope to a class that meets certain conditions.

.`AutoAddScopeBuildItem` Example
[source,java]
----
@BuildStep
AutoAddScopeBuildItem autoAddScope() {
   return AutoAddScopeBuildItem.builder().containsAnnotations(SCHEDULED_NAME, SCHEDULES_NAME) <1>
      .defaultScope(BuiltinScope.SINGLETON) <2>
      .build();
}
----
<1> Find all classes annotated with `@Scheduled`.

> <2> Add `@Singleton`. Classes annotated with a scope are skipped automatically.

Reviewer 3: ```suggestion
<2> Add `@Singleton` as default scope. Classes already annotated with a scope are skipped automatically.
```

   return new AnnotationsTransformerBuildItem(new AnnotationsTransformer() {

      public boolean appliesTo(org.jboss.jandex.AnnotationTarget.Kind kind) {
         return kind == org.jboss.jandex.AnnotationTarget.Kind.CLASS; <1>
      }

      public void transform(TransformationContext context) {
         if (context.getTarget().asClass().name().toString().equals("org.acme.Bar")) {
            context.transform().add(MyInterceptorBinding.class).done(); <2>
         }
      }
    });
}
----
<1> The transformer is only applied to classes.
<2> If the class name equals to `org.acme.Bar` then add `@MyInterceptorBinding`. Don't forget to invoke `Transformation#done()`.

NOTE: Keep in mind that annotation transformers must be produced _before_ the bean discovery starts.

Build steps can query the transformed annotations for a given annotation target via the `TransformedAnnotationsBuildItem`.

.`TransformedAnnotationsBuildItem` Example
[source,java]
----
@BuildStep
void queryAnnotations(TransformedAnnotationsBuildItem transformedAnnotations, BuildProducer<MyBuildItem> myBuildItem) {
   ClassInfo myClazz = ...;
   if (transformedAnnotations.getAnnotations(myClazz).isEmpty()) { <1>

>      myBuildItem.produce(new MyBuildItem(););

Reviewer 3: ```suggestion
     myBuildItem.produce(new MyBuildItem());
```

   }
}

> ----

Reviewer 3: I think it would be nice to add a note saying explicitly that the index is not modified.

Reviewer 2: Hmm, this might be a good addition to the `Metadata Sources` paragraph. Generally stating that extensions alter metamodel but not the index itself.

Author: Metadata Sources is now a separate section and the jandex index is mentioned there. I believe it's clear enough now ;-).

    resourceAnnotations.produce(new ResourceAnnotationBuildItem(DotName.createSimple(MyAnnotation.class.getName())));
}
----

[[build_metadata]]
== Available Build Time Metadata

Any of the above extensions that operates with `BuildExtension.BuildContext` can leverage certain build time metadata that are generated during build.
The built-in keys located in `io.quarkus.arc.processor.BuildExtension.Key` are:

* `ANNOTATION_STORE`
** Contains an `AnnotationStore` that keeps information about all `AnnotationTarget` annotations after application of annotation transformers
* `INJECTION_POINTS`
** `Collection<InjectionPointInfo>` containing all injection points
* `BEANS`
** `Collection<BeanInfo>` containing all beans
* `REMOVED_BEANS`
** `Collection<BeanInfo>` containing all the removed beans; see <<remove_unused_beans>> for more information
* `OBSERVERS`
** `Collection<ObserverInfo>` containing all observers
* `SCOPES`
** `Collection<ScopeInfo>` containing all scopes, including custom ones
* `QUALIFIERS`
** `Map<DotName, ClassInfo>` containing all qualifiers
* `INTERCEPTOR_BINDINGS`
** `Map<DotName, ClassInfo>` containing all interceptor bindings
* `STEREOTYPES`

> ** `Map<DotName, StereotypeInfo>` containing all stereotypes

Reviewer 3: You might or might not want to change it but for this sort of things, you have Asciidoc's description lists: https://asciidoctor.org/docs/asciidoc-writers-guide/#description-lists

Author: Interesting, will use it.

include::./attributes.adoc[]
:numbered:
:sectnums:
:sectnumlevels: 4
:toc:
:toclevels: 2

The ArC container is bootstrapped at build time.
The downside of this approach is that CDI Portable Extensions cannot be supported.
Nevertheless, the functionality can be achieved using the Quarkus-specific extensions API.

The container is bootstrapped in multiple phases.
From a high level perspective these phases go as follows:

1. Initialization
2. Bean discovery 
3. Registration of synthetic components
4. Validation

In the _initialization_ phase the preparatory work is being carried out and custom contexts are registered.
_Bean discovery_ is then the process where the container analyzes all application classes, identifies beans and wires them all together based on the provided metadata.
Subsequently, the extensions can register _synthetic components_.
Attributes of these components are fully contolled by the extensions, i.e. are not derived from an existing class.
Finally, the _deployment is validated_.
For example, the container validates every injection point in the application and fails the build if there is no bean that satisfies the given required type and qualifiers.

> TIP: You can see more information about the bootstrap by enabling additional logging. Simply run the Maven build with `-X,--debug` and grep lines that contain `io.quarkus.arc`. In the <<cdi-reference.adoc#dev-mode,simplified development mode>>, you can use `quarkus.log.category."io.quarkus.arc.processor".level=DEBUG` and two special endpoints are also registered automatically to provide some basic debug info in the JSON format.

Reviewer 4: What's "simplified development mode"? Isn't it just "development mode"?

Author: IDK. copy&paste. I'll fix it.

There are several reasons why a class is not recognized and also several ways to fix it.
In the first step we should identify the _reason_.

[[additional_bean_build_item]]
=== _Reason 1_: Class Is Not discovered 

Quarkus has a <<cdi-reference.adoc#bean_discovery,simplified discovery>>.
It might happen that the class is not part of the application index.
For example, classes from the _runtime module_ of a Quarkus extension are not indexed automatically.

_Solution_: Use the `AdditionalBeanBuildItem`. 
This build item can be used to specify one or more additional classes to be analyzed during the discovery. 
Additional bean classes are transparently added to the application index processed by the container.

.`AdditionalBeanBuildItem` Example
[source,java]
----
@BuildStep
AdditionalBeanBuildItem additionalBeans() {
     return new AdditionalBeanBuildItem(SmallRyeHealthReporter.class, HealthServlet.class)); <1>
}
----
<1> Convenient `AdditionalBeanBuildItem.Builder` can be used for more complex use cases.

Bean classes added via `AdditionalBeanBuildItem` are _removable_ by default.

> If the container considers them <<cdi-reference.adoc#remove_unused_beans,unused>> they are just ignored.

Reviewer 4: ```suggestion
If the container considers them <<cdi-reference.adoc#remove_unused_beans,unused>>, they are just ignored.
```

Bean defining annotations are declared on the class-level and incluce scopes, stereotypes and `@Interceptor`.

_Solution 1_: Use the `AutoAddScopeBuildItem`. This build item can be used to add a scope to a class that meets certain conditions.

.`AutoAddScopeBuildItem` Example
[source,java]
----
@BuildStep
AutoAddScopeBuildItem autoAddScope() {
   return AutoAddScopeBuildItem.builder().containsAnnotations(SCHEDULED_NAME, SCHEDULES_NAME) <1>
      .defaultScope(BuiltinScope.SINGLETON) <2>
      .build();
}
----
<1> Find all classes annotated with `@Scheduled`.
<2> Add `@Singleton`. Classes annotated with a scope are skipped automatically.

_Solution 2_: If you need to process classes annotated with a specific annotation then it's possible to extend the set of bean defining annotations via the `BeanDefiningAnnotationBuildItem`.

.`BeanDefiningAnnotationBuildItem` Example
[source,java]
----
@BuildStep
BeanDefiningAnnotationBuildItem additionalBeanDefiningAnnotation() {
   return new BeanDefiningAnnotationBuildItem(Annotations.GRAPHQL_API); <1>
}
----

> <1> Add `org.eclipse.microprofile.graphql.GraphQLApi` to the set of bean defining annotations.

Reviewer 4: In such case, what's the scope of the bean?

Author: It should be the same as with `AdditionalBeanBuildItem`, i.e. `@Dependent`. I'll add a note...

With an `AnnotationsTransformerBuildItem` it's possible to override the annotations that exist on bean classes.

For example, you might want to add an interceptor binding to a specific bean class.
Here is how to do it: 

.`AnnotationsTransformerBuildItem` Example
[source,java]
----
@BuildStep
AnnotationsTransformerBuildItem transform() {
   return new AnnotationsTransformerBuildItem(new AnnotationsTransformer() {

      public boolean appliesTo(org.jboss.jandex.AnnotationTarget.Kind kind) {
         return kind == org.jboss.jandex.AnnotationTarget.Kind.CLASS; <1>
      }

      public void transform(TransformationContext context) {
         if (context.getTarget().asClass().name().toString().equals("org.acme.Bar")) {
            context.transform().add(MyInterceptorBinding.class).done(); <2>
         }
      }
    });
}
----
<1> The transformer is only applied to classes.
<2> If the class name equals to `org.acme.Bar` then add `@MyInterceptorBinding`. Don't forget to invoke `Transformation#done()`.

NOTE: Keep in mind that annotation transformers must be produced _before_ the bean discovery starts.

> Build steps can query the transformed annotations for a given annotation target via the `TransformedAnnotationsBuildItem`.

Reviewer 4: This build item probably shouldn't be used in some earlier phases? Maybe better to spell it out?

Author: Currently, it's produced during the bean registration phase, after bean discovery but before the synthetic components are registered. In fact, the build items dependency mechanism ensures that a build step that injects this build item is executed after the bean registration phase.

Reviewer 4: Gotcha. It felt weird in the flow of the text, because just a few paragraphs above, you're talking about an `AnnotationsTransformerBuildItem`, and I'm pretty sure you can't have a build step that consumes `TransformedAnnotationsBuildItem` and, at the same time, produces `AnnotationsTransformerBuildItem`.

But perhaps it's just me and that limitation is obvious :-)

.`TransformedAnnotationsBuildItem` Example
[source,java]
----
@BuildStep
void queryAnnotations(TransformedAnnotationsBuildItem transformedAnnotations, BuildProducer<MyBuildItem> myBuildItem) {
   ClassInfo myClazz = ...;
   if (transformedAnnotations.getAnnotations(myClazz).isEmpty()) { <1>
     myBuildItem.produce(new MyBuildItem(););
   }
}
----
<1> `TransformedAnnotationsBuildItem.getAnnotations()` will return a possibly transformed set of annotations.

NOTE: There are other build items specialized on transformation: <<additional_interceptor_bindings>> and <<injection_point_transformation>>.

[[inspect_beans]]
== Use Case - Inspect Beans, Observers and Injection Points

=== _Solution 1_: `BeanDiscoveryFinishedBuildItem`

Consumers of ``BeanDiscoveryFinishedBuildItem`` can easily inspect all class-based beans, observers and injection points registered in the application.
However, synthetic beans and observers are _not included_ because this build item is produced _before_ the synthetic components are registered.

> Additionaly, the bean resolver returned from `BeanDiscoveryFinishedBuildItem#getBeanResolver()` can be used to apply the type-safe resolution rules, e.g. to find out wheter there is a bean that would satisfy certain combination of required type and qualifiers.

Reviewer 4: wheter -> whether

.`BeanDiscoveryFinishedBuildItem` Example
[source,java]
----
@BuildStep
void doSomethingWithNamedBeans(BeanDiscoveryFinishedBuildItem beanDiscovery, BuildProducer<NamedBeansBuildItem> namedBeans) {
   List<BeanInfo> namedBeans = beanDiscovery.beanStream().withName().collect(toList())); <1>
   namedBeans.produce(new NamedBeansBuildItem(namedBeans));
}
----
<1> The resulting list will not contain `@Named` synthetic beans.

=== _Solution 2_: `SynthesisFinishedBuildItem`

Consumers of `SynthesisFinishedBuildItem` can easily inspect all beans, observers and injection points registered in the application. Synthetic beans and observers are included because this build item is produced _after_ the synthetic components are registered.

> Additionaly, the bean resolver returned from `SynthesisFinishedBuildItem#getBeanResolver()` can be used to apply the type-safe resolution rules, e.g. to find out wheter there is a bean that would satisfy certain combination of required type and qualifiers.

Reviewer 4: wheter -> whether

Sometimes it is practical to be able to register a _synthetic bean_.
Bean attributes of a synthetic bean are not derived from a Java class, method or field.
Instead, all the attributes are defined by an extension.
In regular CDI, this could be achieved using the https://docs.jboss.org/cdi/spec/2.0/cdi-spec.html#after_bean_discovery[`AfterBeanDiscovery.addBean()`, window="_blank"] methods.

_Solution_: If you need to register a synthetic bean then use the `SyntheticBeanBuildItem`.

.`SyntheticBeanBuildItem` Example 1
[source,java]
----
@BuildStep
SyntheticBeanBuildItem syntheticBean() {
   return SyntheticBeanBuildItem.configure(String.class)
             .qualifiers(new MyQualifierLiteral())
             .creator(mc -> mc.returnValue(mc.load("foo"))) <1>
             .done();
}
----
<1> Generate the bytecode of the `javax.enterprise.context.spi.Contextual#create(CreationalContext<T>)` implementation.

The output of a bean configurator is recorded as bytecode. 
Therefore, there are some limitations in how a synthetic bean instance is created at runtime. 
You can:

1. Generate the bytecode directly via `ExtendedBeanConfigurator.creator(Consumer<MethodCreator>)`.

> 2. Pass a `io.quarkus.arc.BeanCreator` implementation via `ExtendedBeanConfiguratorio#creator(Class<? extends BeanCreator<U>>)`, and possibly specify some parameters via `ExtendedBeanConfigurator#param()`.

Reviewer 4: `ExtendedBeanConfiguratorio` -> `ExtendedBeanConfigurator`

<2> The bean instance is supplied by a value returned from a recorder method.

It is possible to mark a synthetic bean to be initialized during `RUNTIME_INIT`:

.`RUNTIME_INIT` `SyntheticBeanBuildItem` Example
[source,java]
----
@BuildStep
@Record(RUNTIME_INIT) <1>
SyntheticBeanBuildItem syntheticBean(TestRecorder recorder) {
   return SyntheticBeanBuildItem.configure(Foo.class).scope(Singleton.class)
                .setRuntimeInit() <2>
                .runtimeValue(recorder.createFoo())
                .done();
}
----
<1> The recorder must be executed in the `ExecutionTime.RUNTIME_INIT` phase.
<2> The bean instance is initialized during `RUNTIME_INIT`.

[IMPORTANT]
====
Synthetic beans initialized during `RUNTIME_INIT` must not be accessed during `STATIC_INIT`. `RUNTIME_INIT` build steps that access a runtime-init synthetic bean should consume the `SyntheticBeansRuntimeInitBuildItem`:

[source,java]
----
@BuildStep
@Record(RUNTIME_INIT) 
@Consume(SyntheticBeansRuntimeInitBuildItem.class) <1>
void accessFoo(TestRecorder recorder) {
   recorder.foo(); <2>
}
----
<1> This build step must be executed after `syntheticBean()` completes.

> <2> This recorder method results in an invocation of the `Foo` bean instance.

Reviewer 4: "invocation of the `Foo` bean instance" -- not sure what that means, perhaps you meant "creation" of the bean instance?

Author: Well, "creation" precedes "invocation" ;-). It means that inside the recorder method we do invoke a method upon the `Foo` instance of the bean registered in the snippet above.

Reviewer 4: Ah so it should be "results in an invocation _on"_ the `Foo` bean instance" I think.

====

NOTE: It is also possible to use the `BeanRegistrarBuildItem` and `BeanRegistrationPhaseBuildItem` build items to register a synthetic bean. However, we recommend extension authors to stick with `SyntheticBeanBuildItem` which is more idiomatic for Quarkus.

[[synthetic_observers]]
== Use Case - Synthetic Observers

> Similar to <<synthetic_beans,synthetic beans>> the attributes of a synthetic observer method are not derived from a Java method. Instead, all the attributes are defined by an extension.

Reviewer 4: ```suggestion
Similar to <<synthetic_beans,synthetic beans>>, the attributes of a synthetic observer method are not derived from a Java method. Instead, all the attributes are defined by an extension.
```

=== _Solution 1_: `ObserverRegistrarBuildItem`

> If you need to register a synthetic observer that requires no _build time context_ use the `ObserverRegistrarBuildItem` and leverage the `io.quarkus.arc.processor.ObserverConfigurator` API to build a synthetic observer method definition.

Reviewer 4: ```suggestion
If you need to register a synthetic observer that requires no _build time context_, use the `ObserverRegistrarBuildItem` and leverage the `io.quarkus.arc.processor.ObserverConfigurator` API to build a synthetic observer method definition.
```

== Use Case - I Need to Validate the Deployment

Sometimes extensions need to inspect the beans, observers and injection points, then perform additional validations and fail the build if something is wrong.

=== _Solution 1_: register a `BeanDeploymentValidatorBuildItem`

If you need to validate the deployment with no _build time context_ use the `BeanDeploymentValidatorBuildItem` and leverage the `io.quarkus.arc.processor.BeanDeploymentValidator`.

.`BeanDeploymentValidatorBuildItem` Example
[source,java]
----
@BuildStep
BeanDeploymentValidatorBuildItem beanDeploymentValidator() {
    return new BeanDeploymentValidatorBuildItem(new BeanDeploymentValidator() {
         public void validate(ValidationContext context) {
             for (InjectionPointInfo injectionPoint : context.getInjectionPoints()) {
                 // NOTE: You can't produce a build item from within this callback!
                 System.out.println("Injection point: " + injectionPoint);
             }
         }
    });
}
----

TIP: You can easily filter all registered beans via the convenient `BeanStream` returned from the `ValidationContext.beans()` method.

=== _Solution 2_: `ValidationPhaseBuildItem`

> If an extension needs to produce other build items during the validation it should use the `ValidationPhaseBuildItem` instead.

Reviewer 4: Not for this PR, but this started bugging me in the previous chapters and now I just have to ask: what's the difference between those 2 build items (`BeanDeploymentValidatorBuildItem` and `ValidationPhaseBuildItem`)? It seems to me that the 2nd is stricly more powerful, easier to use and generally feels like a better approach, while the 1st looks like some legacy thing we might wanna deprecate (and possibly even remove) at some point?

Author: Well, you're right that it's a result of API evolution ;-). However, the difference is mentioned in _the first sentence_ of the description of both build items... In the beginning there was the `BeanDeploymentValidatorBuildItem` which is basically a callback that is used inside ArC during bootstrap. This works just fine but the problem is that you can't produce a build item from within this callback, ie. outside a `@BuildStep` invocation (well-defined limitation). And so we had to split the bootstrap into multiple phases and introduce the corresponding `Phase` build items. The downside of this approach is that you need to add the corresponding `BuildProducer` otherwise it could be ignored or processed at the wrong time.

Ideally, we should provide something like `SyntheticBeanBuildItem` for all phases. However, I'm not sure it's worth the effort.

Reviewer 4: Yea I read the first sentence, but it didn't quite occur to me that _this particular difference_ warrants 2 mechanisms to exists to achieve the same thing. It just feels strange.

Author: > It just feels strange.

I agree. One reason was backwards compatibility. The other reason is that neither of these solutions is "good enough". 

We should probably deprecate the `BeanRegistrarBuildItem` because we have the `SyntheticBeanBuildItem` + `BeanDiscoveryFinishedBuildItem`/`SynthesisFinishedBuildItem` combo.

And we should probably deprecate `BeanDeploymentValidatorBuildItem` and `ObserverRegistrarBuildItem` as well.

Author: `BeanRegistrarBuildItem`, `BeanDeploymentValidatorBuildItem` and `ObserverRegistrarBuildItem` are now deprecated. I'm not sure about the `ContextRegistrarBuildItem` - there is actually a chicken-egg problem with scope annotations. I'll try to fix it tomorrow.

Author: `ContextRegistrarBuildItem` deprecated too. We had to introduce a new build item to register the scope annotation - `CustomScopeBuildItem`. The API is a bit cumbersome but I have no other idea how to fix the chicken-egg problem with bean defining annotations.

If an extension needs to produce other build items during the registration it should use the `ContextRegistrationPhaseBuildItem` instead.
The reason is that an injected `BuildProducer` is only valid during a `@BuildStep` method invocation.

IMPORTANT: A build step that consumes the `ContextRegistrationPhaseBuildItem` should always produce a `ContextConfiguratorBuildItem` or at least inject a `BuildProducer` for this build item, otherwise it could be ignored or processed at the wrong time (e.g. after the correct CDI bootstrap phase).

[source,java]
----
@BuildStep
void registerContext(MyConfig config, ContextRegistrationPhaseBuildItem phase,
            BuildProducer<MyBuildItem> myBuildItem, 
            BuildProducer<ContextConfiguratorBuildItem> contexts) {
   if (config.isEnabled()) {
      ContextConfigurator configurator = phase.getContext().configure(TransactionScoped.class).normal().contextClass(TransactionContext.class).done();
      contexts.produce(new ContextConfiguratorBuildItem(configurator));
      myBuildItem.produce(new MyBuildItem());
   }
}
----

=== What if I Need to Know All the Scopes Used in the Application?

> _Solution_: You can inject the `CustomScopeAnnotationsBuildItem` in a build step and use the convenient methods such as `CustomScopeAnnotationsBuildItem.isScopeDeclaredOn()`.

Reviewer 4: Similarly to a previous question -- sounds like this shouldn't be used in some earlier phases?

Author: It can be used before the bootrap actually starts.

[[injection_point_transformation]]
== Use Case - Injection Point Transformation

Every now and then it is handy to be able to change the qualifiers of an injection point programmatically.
You can do just that with `InjectionPointTransformerBuildItem`.
The following sample shows how to apply transformation to injection points with type `Foo` that contain qualifier `MyQualifier`:

.`InjectionPointTransformerBuildItem` Example
[source,java]
----
@BuildStep
InjectionPointTransformerBuildItem transformer() {
    return new InjectionPointTransformerBuildItem(new InjectionPointsTransformer() {

        public boolean appliesTo(Type requiredType) {
            return requiredType.name().equals(DotName.createSimple(Foo.class.getName()));
        }

        public void transform(TransformationContext context) {
            if (context.getQualifiers().stream()
                    .anyMatch(a -> a.name().equals(DotName.createSimple(MyQualifier.class.getName())))) {
                context.transform()
                        .removeAll()
                        .add(DotName.createSimple(MyOtherQualifier.class.getName()))
                        .done();
            }
        }
    });
}
----

> NOTE: In theory, you can use <<annotations_transformer_build_item,an `AnnotationsTransformer`>> to achieve the same goal. However, there are few differences that make `InjectionPointsTransformer` more suitable for this particular task: (1) Annotation transformers are applied before to all classes during bean discovery whereas `InjectionPointsTransformer` is only applied to discovered injection point after bean discovery and (2) with `InjectionPointsTransformer` you don't need to handle various types of injection points (field, parameters of initializer methods, etc.).

Reviewer 4: ```suggestion
NOTE: In theory, you can use <<annotations_transformer_build_item,an `AnnotationsTransformer`>> to achieve the same goal. However, there are few differences that make `InjectionPointsTransformer` more suitable for this particular task: (1) annotation transformers are applied to all classes during bean discovery, whereas `InjectionPointsTransformer` is only applied to discovered injection points after bean discovery; (2) with `InjectionPointsTransformer` you don't need to handle various types of injection points (field, parameters of initializer methods, etc.).
```

== Use Case - Resource Annotations and Injection

> The `ResourceAnnotationBuildItem` can be used to specify resource annotations that make it possible to resolve non-CDI injection points, such as Java EE resources.

Reviewer 4: Jakarta EE perhaps? Not sure what we refer to elsewhere.

    resourceAnnotations.produce(new ResourceAnnotationBuildItem(DotName.createSimple(MyAnnotation.class.getName())));
}
----

[[build_metadata]]
== Available Build Time Metadata

Any of the above extensions that operates with `BuildExtension.BuildContext` can leverage certain build time metadata that are generated during build.
The built-in keys located in `io.quarkus.arc.processor.BuildExtension.Key` are:

* `ANNOTATION_STORE`
** Contains an `AnnotationStore` that keeps information about all `AnnotationTarget` annotations after application of annotation transformers
* `INJECTION_POINTS`
** `Collection<InjectionPointInfo>` containing all injection points
* `BEANS`
** `Collection<BeanInfo>` containing all beans
* `REMOVED_BEANS`
** `Collection<BeanInfo>` containing all the removed beans; see <<remove_unused_beans>> for more information
* `OBSERVERS`
** `Collection<ObserverInfo>` containing all observers
* `SCOPES`
** `Collection<ScopeInfo>` containing all scopes, including custom ones
* `QUALIFIERS`
** `Map<DotName, ClassInfo>` containing all qualifiers
* `INTERCEPTOR_BINDINGS`
** `Map<DotName, ClassInfo>` containing all interceptor bindings
* `STEREOTYPES`

> ** `Map<DotName, ClassInfo>` containing all stereotypes

Reviewer 4: Something I found during the CDI Lite work: this is actually incorrect (and not caught by javac :-) ), the `Map` stored in the context is in fact `Map<DotName, StereotypeInfo>`. Which actually makes more sense.

Author: Ok. This part was CTRL+C and CTRL+V. At least we can fix it right away ;-)

To get hold of these, simply query the extension context object for given key.

> Note that these metadata are made available as build proceeds which means that extensions can only leverage metadata that were build before they are invoked.

Reviewer 4: ```suggestion
Note that these metadata are made available as build proceeds which means that extensions can only leverage metadata that were built before the extensions are invoked.
```

(Two subjects, not clear to which the "they" refers.)

Attributes of these components are fully contolled by the extensions, i.e. are not derived from an existing class.
Finally, the _deployment is validated_.
For example, the container validates every injection point in the application and fails the build if there is no bean that satisfies the given required type and qualifiers.

TIP: You can see more information about the bootstrap by enabling additional logging. Simply run the Maven build with `-X,--debug` and grep lines that contain `io.quarkus.arc`. In the <<cdi-reference.adoc#dev-mode,development mode>>, you can use `quarkus.log.category."io.quarkus.arc.processor".level=DEBUG` and two special endpoints are also registered automatically to provide some basic debug info in the JSON format.

Quarkus build steps can produce and consume various build items and hook into each phase.
In the following sections we will describe all the relevant build items and common scenarios.

[IMPORTANT]
.Metadata Sources
====
Classes and annotations are the primary source of bean-level metadata.
However, extensions can add, remove or transform the metadata at certain stages of the bootstrap.

> Moreover, extenions can also register <<synthetic_beans,synthetic components>>. 

Reviewer 5: s/extenions/extensions

An `UnsatisfiedResolutionException` indicates a problem during <<cdi.adoc#typesafe_resolution,typesafe resolution>>.
Sometimes an injection point cannot be satisifed even if there is a class on the class path that appears to be eligible for injection.
There are several reasons why a class is not recognized and also several ways to fix it.
In the first step we should identify the _reason_.

[[additional_bean_build_item]]
=== _Reason 1_: Class Is Not discovered 

Quarkus has a <<cdi-reference.adoc#bean_discovery,simplified discovery>>.
It might happen that the class is not part of the application index.
For example, classes from the _runtime module_ of a Quarkus extension are not indexed automatically.

_Solution_: Use the `AdditionalBeanBuildItem`. 
This build item can be used to specify one or more additional classes to be analyzed during the discovery. 
Additional bean classes are transparently added to the application index processed by the container.

.`AdditionalBeanBuildItem` Example
[source,java]
----
@BuildStep
AdditionalBeanBuildItem additionalBeans() {
     return new AdditionalBeanBuildItem(SmallRyeHealthReporter.class, HealthServlet.class)); <1>
}
----

> <1> Convenient `AdditionalBeanBuildItem.Builder` can be used for more complex use cases.

Reviewer 5: I think you can drop `convenient` here - it doesn't add anything to the sentence

         }
      }
    });
}
----
<1> The transformer is only applied to classes.
<2> If the class name equals to `org.acme.Bar` then add `@MyInterceptorBinding`. Don't forget to invoke `Transformation#done()`.

NOTE: Keep in mind that annotation transformers must be produced _before_ the bean discovery starts.

Build steps can query the transformed annotations for a given annotation target via the `TransformedAnnotationsBuildItem`.

.`TransformedAnnotationsBuildItem` Example
[source,java]
----
@BuildStep
void queryAnnotations(TransformedAnnotationsBuildItem transformedAnnotations, BuildProducer<MyBuildItem> myBuildItem) {
   ClassInfo myClazz = ...;
   if (transformedAnnotations.getAnnotations(myClazz).isEmpty()) { <1>
     myBuildItem.produce(new MyBuildItem(););
   }
}
----
<1> `TransformedAnnotationsBuildItem.getAnnotations()` will return a possibly transformed set of annotations.

NOTE: There are other build items specialized on transformation: <<additional_interceptor_bindings>> and <<injection_point_transformation>>.

[[inspect_beans]]
== Use Case - Inspect Beans, Observers and Injection Points

=== _Solution 1_: `BeanDiscoveryFinishedBuildItem`

> Consumers of ``BeanDiscoveryFinishedBuildItem`` can easily inspect all class-based beans, observers and injection points registered in the application.

Reviewer 5: Is there a reason why two backticks are used?

.`ObserverRegistrationPhaseBuildItem` Example
[source,java]
----
@BuildStep
void syntheticObserver(ObserverRegistrationPhaseBuildItem observerRegistrationPhase,
            BuildProducer<MyBuildItem> myBuildItem, 
            BuildProducer<ObserverConfiguratorBuildItem> observerConfigurators) {
   observerConfigurators.produce(new ObserverConfiguratorBuildItem(observerRegistrationPhase.getContext().configure().observedType(String.class)
                             .notify(mc -> {
                               // do some gizmo bytecode generation...
                             }).done();
   myBuildItem.produce(new MyBuildItem());
}
----

[[generated_beans]]
== Use Case - I Have a Generated Bean Class

No problem.
You can generate the bytecode of a bean class manually and then all you need to do is to produce a `GeneratedBeanBuildItem` instead of `GeneratedClassBuildItem`.

.`GeneratedBeanBuildItem` Example
[source,java]
----
@BuildStep

> GeneratedBeanBuildItem generatedBean(BuildProducer<GeneratedBeanBuildItem> generatedBeans) {

Reviewer 5: I think you meant to return `void` here

    ClassOutput beansClassOutput = new GeneratedBeanGizmoAdaptor(generatedBeans); <1>
    ClassCreator beanClassCreator = ClassCreator.builder().classOutput(beansClassOutput)
                .className("org.acme.MyBean")
                .build();
    beanClassCreator.addAnnotation(Singleton.class);
    beanClassCreator.close(); <2>
}
----

> <1> A convenient `io.quarkus.arc.deployment.GeneratedBeanGizmoAdaptor` can be used to produce a `GeneratedBeanBuildItem` from Gizmo constructs.

Reviewer 5: I would rephrase as: Using `io.quarkus.arc.deployment.GeneratedBeanGizmoAdaptor` makes it easy to produce `GeneratedBeanBuildItem` from Gizmo constructs
